From f08fa56dbfbde9bebb8caa4087fffe1040c1b52d Mon Sep 17 00:00:00 2001
From: Arulselvan M <arulselvan.m@intel.com>
Date: Thu, 30 May 2013 20:27:18 +0530
Subject: [PORT FROM MAIN][HDMI][EDID]Parse channel info and use it for sink open

BZ: 60825

The APM is updated to return direct profile only
for multichannel contents when HDMI is connected. This
is done to accomodate the dynamic channel count read
from EDID.

Change-Id: I6bbf6a9d5178788eab5bb6b6668db87fc0113cb9
Orig-Change-Id: Ie0c295a1ca79c29c7fe7abb71d255cc3bd9f70b5
Signed-off-by: jerome <jerome.anand@intel.com>
Signed-off-by: Arulselvan M <arulselvan.m@intel.com>
---
 audio/AudioPolicyManagerBase.cpp |   23 ++++++++++++++++++++++-
 1 files changed, 22 insertions(+), 1 deletions(-)

diff --git a/audio/AudioPolicyManagerBase.cpp b/audio/AudioPolicyManagerBase.cpp
index e2ae86e..9202f8e 100644
--- a/audio/AudioPolicyManagerBase.cpp
+++ b/audio/AudioPolicyManagerBase.cpp
@@ -506,7 +506,14 @@ AudioPolicyManagerBase::IOProfile *AudioPolicyManagerBase::getProfileForDirectOu
         }
         for (size_t j = 0; j < mHwModules[i]->mOutputProfiles.size(); j++) {
            IOProfile *profile = mHwModules[i]->mOutputProfiles[j];
-           if (profile->isCompatibleProfile(device, samplingRate, format,
+           if((channelMask <= AUDIO_CHANNEL_OUT_STEREO) &&
+                    (device == AudioSystem::DEVICE_OUT_AUX_DIGITAL)){
+              /*Direct output is selected only for multichannel content over HDMI
+                NOTE - if stereo contents needs direct thread over HDMI; this condition
+                       needs to be removed*/
+              ALOGI("make direct output unavaiable for stereo clips");
+              return 0;
+           }else if (profile->isCompatibleProfile(device, samplingRate, format,
                                            channelMask,
                                            AUDIO_OUTPUT_FLAG_DIRECT)) {
                if (mAvailableOutputDevices & profile->mSupportedDevices) {
@@ -716,6 +723,20 @@ status_t AudioPolicyManagerBase::startOutput(audio_io_handle_t output,
        mIsDirectOutputActive = true;
     }
 
+
+    if (outputDesc->mFlags & AUDIO_OUTPUT_FLAG_COMPRESS_OFFLOAD) {
+
+       // Informs primary HAL that a compressed output will be started
+       AudioParameter param;
+       param.addInt(String8(AudioParameter::keyStreamFlags), AUDIO_OUTPUT_FLAG_COMPRESS_OFFLOAD);
+       mpClientInterface->setParameters(0, param.toString(), 0);
+    }
+    if((outputDesc->mFlags & AudioSystem::OUTPUT_FLAG_DIRECT) && (mAvailableOutputDevices & AudioSystem::DEVICE_OUT_AUX_DIGITAL)
+        && (outputDesc->mChannelMask != AUDIO_CHANNEL_OUT_STEREO)){
+       ALOGD("startoutput() Direct thread is active for 0x%x channels",outputDesc->mChannelMask);
+       mIsDirectOutputActive = true;
+    }
+
     if (outputDesc->mRefCount[stream] == 1) {
         audio_devices_t newDevice = getNewDevice(output, false /*fromCache*/);
         routing_strategy strategy = getStrategy(stream);
-- 
1.7.4.1

